8333725: Compiler Prototyping for Deconstructors

biboudis opened the pull request:
Work-in-progress prototyping of deconstructors.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Warnings
&nbsp;⚠️ Patch contains a binary file (src/java.desktop/share/classes/javax/swing/beaninfo/images/JAppletColor16.gif)
&nbsp;⚠️ Patch contains a binary file (src/java.desktop/share/classes/javax/swing/beaninfo/images/JAppletColor32.gif)
&nbsp;⚠️ Patch contains a binary file (src/java.desktop/share/classes/javax/swing/beaninfo/images/JAppletMono16.gif)
&nbsp;⚠️ Patch contains a binary file (src/java.desktop/share/classes/javax/swing/beaninfo/images/JAppletMono32.gif)
&nbsp;⚠️ Patch contains a binary file (src/java.desktop/share/classes/javax/swing/doc-files/JRootPane-1.gif)
&nbsp;⚠️ Patch contains a binary file (test/jdk/javax/rmi/ssl/keystore)
&nbsp;⚠️ Patch contains a binary file (test/jdk/javax/rmi/ssl/truststore)
&nbsp;⚠️ Patch contains a binary file (test/jdk/jdk/internal/loader/URLClassPath/testclasses.jar)
&nbsp;⚠️ Patch contains a binary file (test/jdk/sun/net/www/protocol/https/HttpsClient/dnsstore)
&nbsp;⚠️ Patch contains a binary file (test/jdk/sun/net/www/protocol/https/HttpsClient/ipstore)

### Issue
 * [JDK-8333725](https://bugs.openjdk.org/browse/JDK-8333725): Compiler Prototyping for Deconstructors (**Enhancement** - P4)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/amber.git pull/94/head:pull/94` \
`$ git checkout pull/94`

Update a local copy of the PR: \
`$ git checkout pull/94` \
`$ git pull https://git.openjdk.org/amber.git pull/94/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 94`

View PR using the GUI difftool: \
`$ git pr show -t 94`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/amber/pull/94.diff">https://git.openjdk.org/amber/pull/94.diff</a>

</details>

vicente-romero-oracle commented on src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Attr.java:
```diff
@@ -4286,6 +4366,70 @@ public void visitRecordPattern(JCRecordPattern tree) {
         matchBindings = new MatchBindings(outBindings.toList(), List.nil());
     }
 
+    // todo: follow the protocol in Resolve::selectBest
```
this method I think belongs in Resolve

vicente-romero-oracle commented on src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Lower.java:
```diff
@@ -2779,6 +2779,32 @@ public void visitMethodDef(JCMethodDecl tree) {
 
+        if (tree.sym.isPattern()) {
+            tree.sym.flags_field  |= STATIC;
+            tree.sym.flags_field  |= SYNTHETIC;
```
we are making synthetic an AST that actually appears in the source?

vicente-romero-oracle commented on src/jdk.compiler/share/classes/com/sun/tools/javac/parser/JavacParser.java:
```diff
@@ -198,6 +199,7 @@ protected JavacParser(ParserFactory fac,
         this.allowYieldStatement = Feature.SWITCH_EXPRESSION.allowedInSource(source);
         this.allowRecords = Feature.RECORDS.allowedInSource(source);
         this.allowSealedTypes = Feature.SEALED_CLASSES.allowedInSource(source);
+        this.allowPatternDeclarations = Feature.PATTERN_DECLARATIONS.allowedInSource(source);
```
preview mode should be enable too

vicente-romero-oracle commented on test/test/Point.class:
this one was uploaded by mistake I guess

biboudis committed:
8333725: Compiler Prototyping for Deconstructors

Co-authored-by: Angelos Bimpoudis <angelos.bimpoudis@oracle.com>
Co-authored-by: Jan Lahoda <jan.lahoda@oracle.com>
Co-authored-by: Vicente Romero <vicente.romero@oracle.com>

vicente-romero-oracle commented on src/jdk.compiler/share/classes/com/sun/tools/javac/jvm/Items.java:
```diff
@@ -455,7 +455,7 @@ void store() {
 
         Item invoke() {
-            MethodType mtype = (MethodType)member.erasure(types);
+            MethodType mtype = (MethodType)member.externalType(types);
```
not sure about this change

biboudis replied:
The thing is that when the time comes to Generate a var def. For example `/*synthetic*/ final Object patt4$temp = ExampleStringTest$StringOps$Split.Split($b$0)`, the result needs to be `stack(object)` instead of `void` for the deconstructor call. The method type of split needs to be `(ExampleStringTest.StringOps.Split)java.lang.Object`. How can I change that earlier?

vicente-romero-oracle commented on src/jdk.compiler/share/classes/com/sun/tools/javac/jvm/PoolWriter.java:
```diff
@@ -201,7 +202,7 @@ int putName(Name name) {
      */
     int putNameAndType(Symbol s) {
-        return pool.writeIfNeeded(new NameAndType(s.name, descriptorType(s)));
+        return pool.writeIfNeeded(new NameAndType(s instanceof MethodSymbol ms ? ms.externalName(types) : s.name, descriptorType(s)));
```
same not sure why you need this change

biboudis committed:
Fix translation of dtors with primitives and array types

biboudis committed:
Merge remote-tracking branch 'refs/remotes/upstream-amber/master' into matchers-experiment

# Conflicts:
#	src/java.base/share/classes/java/lang/Class.java

biboudis committed:
Restore JTreg flag

biboudis commented on src/java.base/share/classes/java/lang/Class.java:
```diff
@@ -2286,6 +2302,277 @@ public Constructor<?>[] getConstructors() throws SecurityException {
+            ClassModel cm = ClassFile.of().parse(bytes);
+            for (MethodModel mm : cm.methods()) {
+                PatternAttribute pa = null;
+                for (Attribute<?> attribute : mm.attributes()) {
```
Replacing, preview can now be used.

biboudis committed:
Address review 1/n

biboudis committed:
Allow deeper reflection for getDeclaredDeconstructors

biboudis committed:
Replace with getSystemResourceAsStream

lahodaj committed:
Attempting to fix ElementKind for variables inside a deconstructor/pattern - only the bindings should presumably have ElementKind.PATTERN_BINDING, the other variables should have their 'normal' kinds(?) (#5)

biboudis committed:
Fix NPE for null ClassLoader

biboudis committed:
Merge remote-tracking branch 'origin-amber/matchers-experiment' into matchers-experiment

biboudis committed:
Fix mangling

biboudis committed:
Fix private

biboudis committed:
Fix tests after depending on java.base/sun.invoke.util

biboudis committed:
Add test for NestedPatternDeclarations

biboudis committed:
Records have an implicit deconstructor declaration that cannot be explicitly specified

biboudis committed:
Adjust mangling

biboudis committed:
Adjusts pattern overloading

biboudis committed:
Align overloading of patterns in classes with record classes

biboudis committed:
Merge remote-tracking branch 'refs/remotes/upstream-amber/master' into matchers-experiment

# Conflicts:
#	src/java.base/share/classes/jdk/internal/classfile/impl/BoundAttribute.java
#	src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Lower.java
#	src/jdk.compiler/share/classes/com/sun/tools/javac/comp/TransPatterns.java
#	test/jdk/jdk/classfile/StackMapsTest.java

biboudis committed:
Clean-up whitespaces

biboudis committed:
Support running in non--parameters-mode

lahodaj committed:
Making annotation processing work with patterns. (#6)

biboudis committed:
Fix annotation processing with matchers

biboudis committed:
Fix annotation processing with matchers (2)

biboudis committed:
Update Indify

biboudis commented:
Keep it, keep it, keep it.

biboudis committed:
Fix exhaustiveness bug

lahodaj committed:
A crude sketch for Lookup.unreflectDeconstructor (#7)

* A crude sketch of unreflect for deconstructors.

* Reuse the same mangling everywhere.

* Cleanup.

biboudis committed:
Fix uniqueness of patterns

biboudis committed:
Fix resolved pattern declaration resolution check for null

biboudis committed:
Resolve unreflected deconstructor to work with invokeExact

Co-authored-by: Viktor Klang <viktor.klang@oracle.com>

viktorklang-ora committed:
Switching to Carriers.boxedComponentValueArray (#8)

* Switching to Carriers.boxedComponentValueArray

* Fixing UnreflectPattern

biboudis committed:
Introduce Deconstructor <: MemberPattern <: Executable (#9)

* Introduce Deconstructor <: MemberPattern <: Executable

* Fix MemberPattern.setAccessible

* Retain signature of unreflectDeconstructor

biboudis committed:
Add instanceof statement (#10)

biboudis committed:
Invoke Patterns through PatternBootstraps

Rick-24 commented:
您好，您的来信我已收到，我会尽快处理。      祝好！

biboudis commented:
Definitely keep it open :D

biboudis committed:
Remove VM preliminary work

biboudis committed:
Merge remote-tracking branch 'upstream-amber/master' into matchers-experiment

# Conflicts:
#	src/java.base/share/classes/java/lang/classfile/AccessFlags.java
#	src/java.base/share/classes/java/lang/classfile/Attribute.java
#	src/java.base/share/classes/java/lang/classfile/AttributedElement.java
#	src/java.base/share/classes/java/lang/classfile/MethodElement.java
#	src/java.base/share/classes/java/lang/reflect/ReflectAccess.java
#	src/java.base/share/classes/jdk/internal/classfile/impl/ClassRemapperImpl.java
#	src/java.base/share/classes/jdk/internal/classfile/impl/UnboundAttribute.java
#	src/java.base/share/native/libjava/Class.c
#	src/java.compiler/share/classes/javax/lang/model/element/Modifier.java
#	src/java.compiler/share/classes/javax/lang/model/util/AbstractElementVisitorPreview.java
#	src/java.compiler/share/classes/javax/lang/model/util/ElementKindVisitor6.java
#	src/jdk.compiler/share/classes/com/sun/tools/javac/jvm/ClassReader.java
#	test/jdk/jdk/classfile/StackMapsTest.java

biboudis committed:
Merge remote-tracking branch 'upstream-amber/master' into matchers-experiment

# Conflicts:
#	src/jdk.compiler/share/classes/com/sun/tools/javac/code/Source.java

biboudis committed:
Fix imports

biboudis committed:
Fix crashes

- Fix crash when assigning to bindings
- Disallow pattern no matching on all control paths

Co-authored-by: Jan Lahoda <jan.lahoda@oracle.com>

biboudis committed:
Fix target checking for ElementType.DECONSTRUCTOR/METHOD when applied to deconstructors

biboudis committed:
Fix examples

biboudis committed:
Merge branch 'master' into matchers-experiment

# Conflicts:
#	test/jdk/jdk/classfile/StackMapsTest.java

biboudis committed:
Merge remote-tracking branch 'upstream-amber/master' into matchers-experiment

# Conflicts:
#	src/java.base/share/classes/java/lang/reflect/Constructor.java

biboudis committed:
Merge remote-tracking branch 'upstream-amber/master' into matchers-experiment

# Conflicts:
#	src/java.base/share/classes/java/lang/Class.java
#	src/jdk.compiler/share/classes/com/sun/tools/javac/code/Source.java

lahodaj committed:
Initial support for pattern in ct.sym. (#11)

biboudis committed:
Merge remote-tracking branch 'upstream-amber/master' into matchers-experiment

# Conflicts:
#	test/jdk/jdk/classfile/helpers/RebuildingTransformation.java

biboudis commented:
Most definitely keep it open :D

biboudis committed:
Merge remote-tracking branch 'upstream-amber/master' into matchers-experiment

# Conflicts:
#	src/java.base/share/classes/java/lang/Class.java
#	test/jdk/jdk/classfile/helpers/RebuildingTransformation.java

biboudis committed:
Adjust Overloading (#12)

lahodaj committed:
Separate PatternType from MethodType (#13)

* An experiment on PatternType.

* Fixing test.

* Fixing build.

* Fixing tests.

* Cleanup.

biboudis committed:
Merge branch 'master' into matchers-experiment

# Conflicts:
#	src/java.base/share/classes/java/lang/classfile/Attributes.java

biboudis committed:
Fix in Flow

biboudis committed:
Fix whitespaces

biboudis committed:
Fix instanceof statement reporting

lahodaj committed:
Properly reading and applying Signature attribute/type for patterns (#14)

biboudis committed:
Fix partial order in pattern overloading

biboudis committed:
Add the missing case

biboudis committed:
Improve error reporting around overloading

lahodaj committed:
Attempting to add match-failed + partial deconstructors. No support for partial deconstructor exhaustiveness groups yet. (#15)

* Attempting to add match-failed + partial deconstructors. No support for partial deconstructor exhaustiveness groups yet.

* Error recovery.

lahodaj committed:
Addding forgotten file. (#16)

biboudis committed:
Fixing tests. (#17)

Co-authored-by: Jan Lahoda <jan.lahoda@oracle.com>

biboudis committed:
Erase pattern signature for deconstructors (implicit or explicit) (#19)

* Erase pattern signature for records

* Erase pattern signature for deconstructors

biboudis committed:
Merge branch 'master-amber' into matchers-experiment

# Conflicts:
#	src/jdk.compiler/share/classes/com/sun/tools/javac/jvm/ClassReader.java

biboudis committed:
Clean whitespaces

biboudis committed:
Fix separate compilation issue for generic patterns in ClassReader and mangled name calculation

biboudis committed:
Whitespaces

biboudis committed:
Avoid null in PatternType

lahodaj committed:
Precomputing accessors and Carrier constructor for the synthetic pattern. (#21)

lahodaj committed:
First round of instance patterns. No patterns with separate instance and match candidate yet. (#20)

* First round of instance patterns. No patterns with separate instance and match candidate yet.

* A very (VERY!) basic support for instance pattern with split receiver and match candidate.

* Align calling convention between instance patterns and deconstructors

* Fix adjustment of `syntheticPattern` computation

---------

Co-authored-by: Angelos Bimpoudis <angelos.bimpoudis@oracle.com>

biboudis committed:
Handle MatchCandidate properly (#22)

biboudis committed:
Pretty

biboudis committed:
Merge branch 'master-amber' into matchers-experiment

biboudis committed:
Restore deleted import

biboudis committed:
Instance patterns represented by instance methods in the translation (#23)

* Instance patterns represented by instance methods in the translation

* Second Approach: rely on the mangled name

biboudis committed:
Add static patterns (#24)

biboudis committed:
Whitespaces

biboudis committed:
Merge branch 'master-amber' into matchers-experiment

biboudis committed:
Homogenise translation. All pattern methods use one argument

biboudis commented:
/open

biboudis committed:
Merge branch 'master-amber' into matchers-experiment

lahodaj committed:
An attempt to parse instance patterns more correctly.

lahodaj committed:
Progress on parsing instance patterns in instanceof

biboudis committed:
Add extra parameter for initialized carrier (#25)

lahodaj committed:
Improving error recovery.

lahodaj committed:
Merge remote-tracking branch 'upstream-amber/master' into matchers-experiment

lahodaj committed:
Merge remote-tracking branch 'biboudis-amber/matchers-experiment' into matchers-experiment

lahodaj committed:
Using ParameterableDescription.

biboudis committed:
Merge pull request #26 from lahodaj/matchers-experiment

Merging amber/master into matchers-experiment

biboudis committed:
Whitespaces

biboudis committed:
Fix parameterableAttributes order for patternDescription

biboudis committed:
Merge branch 'master-amber' into matchers-experiment-after-newaccessflags

# Conflicts:
#	src/java.base/share/classes/java/lang/reflect/AccessFlag.java

biboudis committed:
Fix AttributeWriter and ClassFile after new AccessFlags

liach committed:
Fix matcher and access flags

biboudis committed:
Merge pull request #27 from liach/fix/matcher-af

Fix matcher and access flags

biboudis committed:
Merge branch 'master' into matchers-experiment

# Conflicts:
#	src/jdk.compiler/share/classes/com/sun/tools/javac/code/Preview.java
#	src/jdk.compiler/share/classes/com/sun/tools/javac/parser/JavacParser.java

biboudis committed:
Update ListModuleDeps

biboudis committed:
Attempt 1

biboudis committed:
Merge pull request #28 from biboudis/matchers-experiment-fix-windows-build

Fix Windows Build

biboudis committed:
Merge branch 'master-amber' into matchers-experiment

# Conflicts:
#	src/jdk.compiler/share/classes/com/sun/tools/javac/code/Preview.java

kevinb9n committed:
write more tests for basic scenarios

kevinb9n committed:
more basic basic tests

kevinb9n committed:
Produce Dtor<T> instead of Dtor<?>

kevinb9n committed:
add PatternMember.getCandidateType(); also remove `slot` for now (can add it back when it serves a purpose)

kevinb9n committed:
rename MemberPattern -> PatternMember

kevinb9n committed:
rename MemberPattern -> PatternMember

kevinb9n committed:
move PatternMember.invoke to Deconstructor.invoke (a 2-arg version might come back)

kevinb9n committed:
hardcode checked exception types to empty array; move getName impl down to Dtor

kevinb9n committed:
populate Parameters almost correctly (missing generic types); PatternBinding is still there too for now

kevinb9n committed:
Merge branch 'matchers-experiment' into matchers-experiment-reflection-first

# Conflicts:
#	src/java.base/share/classes/java/lang/Class.java
#	src/java.base/share/classes/java/lang/reflect/Deconstructor.java
#	src/java.base/share/classes/java/lang/reflect/PatternMember.java

kevinb9n committed:
merge in updates from biboudis/matchers-exp and fix problems

kevinb9n committed:
mostly javadoc and minor touchups; made tryMatch actually require a T

biboudis commented:
/keepalive

biboudis committed:
Merge branch 'master-amber' into matchers-experiment

# Conflicts:
#	make/langtools/src/classes/build/tools/symbolgenerator/CreateSymbols.java

lahodaj committed:
Improving Carriers, adjusting pattern translation to the new Carriers.

lahodaj committed:
Fixing test.

lahodaj committed:
Cleanup.

kevinb9n committed:
add tests demonstrating inner class behavior

kevinb9n committed:
Oh right, getCandidateType needs to exist in three forms

lahodaj committed:
The reflection does not need to go through Carriers.

lahodaj committed:
Cleanup - removing unnecessary throws.

lahodaj committed:
Simplifying reflection, as suggested

kevinb9n committed:
Goodbye to the PatternMember type parameter <T>

lahodaj committed:
Merge branch 'matchers-experiment' into matchers-experiment-instance-pattern-parsing

lahodaj committed:
Cleanup.

lahodaj committed:
Fixing tests.

lahodaj committed:
Fixing a crash when javac is highlighting a deconstructor

biboudis committed:
Merge pull request #32 from lahodaj/fixing-crash-deconstructor

Fixing a crash when javac is highlighting a deconstructor

biboudis committed:
Fix static patterns

lahodaj committed:
Fixing tests.

biboudis committed:
Merge pull request #30 from lahodaj/matchers-experiment-new-carriers

Carriers update for matchers-experiment

biboudis committed:
Adjust overloading and parsing

lahodaj committed:
Merge branch 'matchers-experiment' into matchers-experiment-instance-pattern-parsing

lahodaj committed:
Merge remote-tracking branch 'lahodaj-amber/matchers-experiment-instance-pattern-parsing' into matchers-experiment-instance-pattern-parsing

lahodaj committed:
Fixing some pattern/expression disambiguation and some tests.

lahodaj committed:
Fixing tests.

biboudis committed:
Merge branch 'matchers-experiment-instance-pattern-parsing' into matchers-experiment-instance-pattern-parsing-overloading

lahodaj committed:
Fixing DeduplicationTest.

lahodaj committed:
Merge branch 'matchers-experiment' into matchers-experiment-instance-pattern-parsing

lahodaj committed:
Fixing build.

biboudis committed:
Merge pull request #31 from lahodaj/matchers-experiment-instance-pattern-parsing

An attempt to parse instance patterns more correctly.

biboudis committed:
Merge branch 'matchers-experiment' into matchers-experiment-instance-pattern-parsing-overloading

# Conflicts:
#	src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Attr.java

biboudis committed:
Merge pull request #33 from biboudis/matchers-experiment-instance-pattern-parsing-overloading

Adjusting overloading + translation

biboudis committed:
Merge branch 'matchers-experiment-reflection-first' into matchers-experiment

biboudis committed:
Drop generic parameter from Deconstructor
